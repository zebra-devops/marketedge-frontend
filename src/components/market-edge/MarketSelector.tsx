import React, { useState, useEffect } from 'react';
import { ChevronDownIcon, PlusIcon } from '@heroicons/react/24/outline';
import { Market } from '@/types/market-edge';
import { marketEdgeAPI } from '@/services/market-edge-api';

interface MarketSelectorProps {
  selectedMarket?: Market;
  onMarketSelect: (market: Market) => void;
  onCreateMarket?: () => void;
  className?: string;
}

export const MarketSelector: React.FC<MarketSelectorProps> = ({
  selectedMarket,
  onMarketSelect,
  onCreateMarket,
  className = ''
}) => {
  const [markets, setMarkets] = useState<Market[]>([]);
  const [isOpen, setIsOpen] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  const [searchQuery, setSearchQuery] = useState('');
  const [error, setError] = useState<string | null>(null);
  const [retryCount, setRetryCount] = useState(0);

  useEffect(() => {
    loadMarkets();
  }, []);

  const loadMarkets = async (showLoading = true) => {
    try {
      if (showLoading) {
        setIsLoading(true);
      }
      console.log('Loading markets...');
      const marketData = await marketEdgeAPI.getMarkets();
      setMarkets(marketData || []);
      setError(null);
      setRetryCount(0); // Reset retry count on success
      console.log('Markets loaded successfully:', marketData?.length || 0);
    } catch (err: any) {
      console.error('Failed to load markets:', err);
      
      // Enhanced error handling with specific messages
      let errorMessage = 'Failed to load markets. Please try again.';
      
      if (err?.response?.status === 401) {
        errorMessage = 'Authentication required. Please log in to access markets.';
      } else if (err?.response?.status === 403) {
        errorMessage = 'Access denied. You may not have permission to view markets.';
      } else if (err?.response?.status === 404) {
        errorMessage = 'Markets service not found. Please contact support.';
      } else if (err?.response?.status >= 500) {
        errorMessage = 'Server error occurred. Please try again in a moment.';
      } else if (err?.message?.includes('Network Error')) {
        errorMessage = 'Network connection error. Please check your connection.';
      } else if (err?.message?.includes('timeout')) {
        errorMessage = 'Request timed out. Please try again.';
      } else if (err?.message) {
        errorMessage = err.message;
      }
      
      setError(errorMessage);
      setMarkets([]); // Ensure markets is always an array
      setRetryCount(prev => prev + 1);
    } finally {
      if (showLoading) {
        setIsLoading(false);
      }
    }
  };

  const filteredMarkets = (markets || []).filter(market =>
    market.name.toLowerCase().includes(searchQuery.toLowerCase())
  );

  const handleMarketSelect = (market: Market) => {
    onMarketSelect(market);
    setIsOpen(false);
    setSearchQuery('');
  };

  return (
    <div className={`relative ${className}`}>
      {/* Selected Market Display */}
      <button
        onClick={() => setIsOpen(!isOpen)}
        className="w-full flex items-center justify-between px-4 py-2 bg-white border border-gray-300 rounded-lg shadow-sm hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
        disabled={isLoading}
      >
        <div className="flex items-center">
          <div className="w-3 h-3 bg-green-500 rounded-full mr-3"></div>
          <span className="text-sm font-medium text-gray-900">
            {selectedMarket ? selectedMarket.name : 'Select Market'}
          </span>
        </div>
        <ChevronDownIcon 
          className={`w-5 h-5 text-gray-400 transition-transform ${isOpen ? 'rotate-180' : ''}`} 
        />
      </button>

      {/* Dropdown */}
      {isOpen && (
        <div className="absolute z-50 w-full mt-1 bg-white border border-gray-200 rounded-lg shadow-lg max-h-80 overflow-hidden">
          {/* Search Input */}
          <div className="p-3 border-b border-gray-200">
            <input
              type="text"
              placeholder="Search markets..."
              value={searchQuery}
              onChange={(e) => setSearchQuery(e.target.value)}
              className="w-full px-3 py-2 border border-gray-300 rounded-md text-sm focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
              autoFocus
            />
          </div>

          {/* Markets List */}
          <div className="max-h-60 overflow-y-auto">
            {error ? (
              <div className="p-4 text-center text-red-600 text-sm">
                <div className="mb-2">{error}</div>
                {retryCount > 0 && (
                  <div className="text-xs text-gray-500 mb-2">
                    Retry attempt: {retryCount}
                  </div>
                )}
                <button
                  onClick={() => loadMarkets()}
                  disabled={isLoading}
                  className="block mx-auto mt-2 text-blue-600 hover:text-blue-800 underline disabled:opacity-50 disabled:cursor-not-allowed"
                >
                  {isLoading ? 'Retrying...' : 'Try Again'}
                </button>
                {retryCount >= 3 && (
                  <div className="mt-2 text-xs text-gray-500">
                    Having trouble? Try refreshing the page or contact support.
                  </div>
                )}
              </div>
            ) : isLoading ? (
              <div className="p-4 text-center text-gray-500 text-sm">
                Loading markets...
              </div>
            ) : filteredMarkets.length === 0 ? (
              <div className="p-4 text-center text-gray-500 text-sm">
                {searchQuery ? 'No markets found' : 'No markets available'}
              </div>
            ) : (
              filteredMarkets.map((market) => (
                <button
                  key={market.id}
                  onClick={() => handleMarketSelect(market)}
                  className="w-full px-4 py-3 text-left hover:bg-gray-50 focus:outline-none focus:bg-gray-50 transition-colors"
                >
                  <div className="flex items-center justify-between">
                    <div>
                      <div className="text-sm font-medium text-gray-900">
                        {market.name}
                      </div>
                      <div className="text-xs text-gray-500 mt-1">
                        {market.competitor_count} competitors â€¢ 
                        {market.is_active ? ' Active' : ' Inactive'}
                      </div>
                    </div>
                    {selectedMarket?.id === market.id && (
                      <div className="w-2 h-2 bg-blue-600 rounded-full"></div>
                    )}
                  </div>
                </button>
              ))
            )}
          </div>

          {/* Create New Market */}
          {onCreateMarket && (
            <div className="border-t border-gray-200">
              <button
                onClick={() => {
                  onCreateMarket();
                  setIsOpen(false);
                }}
                className="w-full px-4 py-3 text-left hover:bg-gray-50 focus:outline-none focus:bg-gray-50 transition-colors flex items-center text-blue-600"
              >
                <PlusIcon className="w-4 h-4 mr-2" />
                <span className="text-sm font-medium">Create New Market</span>
              </button>
            </div>
          )}
        </div>
      )}

      {/* Overlay */}
      {isOpen && (
        <div
          className="fixed inset-0 z-40"
          onClick={() => setIsOpen(false)}
        />
      )}
    </div>
  );
};