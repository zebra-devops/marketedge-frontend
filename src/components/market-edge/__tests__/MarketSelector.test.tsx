/**
 * Test file for MarketSelector component
 * 
 * Demonstrates multi-tenant testing patterns with industry-specific scenarios
 */

import React from 'react'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { 
  renderWithProviders,
  renderForIndustry,
  mockApiEndpoints,
  testAccessibility,
  industryTestScenarios,
  generateTestData,
  waitForLoadingToFinish,
} from '../../../utils/test-utils'
import { MarketSelector } from '../MarketSelector'

// Mock the MarketSelector component for testing
jest.mock('../MarketSelector', () => ({
  MarketSelector: ({ onMarketChange, currentMarket, industry }: any) => (
    <div data-testid="market-selector">
      <select
        value={currentMarket || ''}
        onChange={(e) => onMarketChange?.(e.target.value)}
        data-testid="market-select"
      >
        <option value="">Select Market</option>
        {industry === 'cinema' && (
          <>
            <option value="cinema-downtown">Downtown Cinema</option>
            <option value="cinema-mall">Mall Cinema</option>
            <option value="cinema-suburban">Suburban Cinema</option>
          </>
        )}
        {industry === 'hotel' && (
          <>
            <option value="hotel-luxury">Luxury Hotels</option>
            <option value="hotel-business">Business Hotels</option>
            <option value="hotel-budget">Budget Hotels</option>
          </>
        )}
        {industry === 'gym' && (
          <>
            <option value="gym-premium">Premium Fitness</option>
            <option value="gym-community">Community Gyms</option>
            <option value="gym-24h">24/7 Gyms</option>
          </>
        )}
        {(!industry || industry === 'general') && (
          <>
            <option value="general-local">Local Market</option>
            <option value="general-regional">Regional Market</option>
            <option value="general-national">National Market</option>
          </>
        )}
      </select>
      <div data-testid="market-info">
        Current Industry: {industry || 'general'}
      </div>
    </div>
  ),
}))

describe('MarketSelector Component', () => {
  
  describe('Basic functionality', () => {
    it('renders with default props', () => {
      renderWithProviders(<MarketSelector />)
      
      expect(screen.getByTestId('market-selector')).toBeInTheDocument()
      expect(screen.getByTestId('market-select')).toBeInTheDocument()
    })
    
    it('displays current market selection', () => {
      renderWithProviders(
        <MarketSelector currentMarket="general-local" />
      )
      
      const select = screen.getByTestId('market-select') as HTMLSelectElement
      expect(select.value).toBe('general-local')
    })
    
    it('calls onMarketChange when selection changes', async () => {
      const onMarketChange = jest.fn()
      const { user } = renderWithProviders(
        <MarketSelector onMarketChange={onMarketChange} />
      )
      
      const select = screen.getByTestId('market-select')
      await user.selectOptions(select, 'general-regional')
      
      expect(onMarketChange).toHaveBeenCalledWith('general-regional')
    })
  })
  
  describe('Industry-specific behavior', () => {
    it('shows cinema-specific markets for cinema industry', () => {
      renderForIndustry(
        <MarketSelector />,
        'cinema'
      )
      
      expect(screen.getByText('Downtown Cinema')).toBeInTheDocument()
      expect(screen.getByText('Mall Cinema')).toBeInTheDocument()
      expect(screen.getByText('Suburban Cinema')).toBeInTheDocument()
      expect(screen.getByText('Current Industry: cinema')).toBeInTheDocument()
    })
    
    it('shows hotel-specific markets for hotel industry', () => {
      renderForIndustry(
        <MarketSelector />,
        'hotel'
      )
      
      expect(screen.getByText('Luxury Hotels')).toBeInTheDocument()
      expect(screen.getByText('Business Hotels')).toBeInTheDocument()
      expect(screen.getByText('Budget Hotels')).toBeInTheDocument()
      expect(screen.getByText('Current Industry: hotel')).toBeInTheDocument()
    })
    
    it('shows gym-specific markets for gym industry', () => {
      renderForIndustry(
        <MarketSelector />,
        'gym'
      )
      
      expect(screen.getByText('Premium Fitness')).toBeInTheDocument()
      expect(screen.getByText('Community Gyms')).toBeInTheDocument()
      expect(screen.getByText('24/7 Gyms')).toBeInTheDocument()
      expect(screen.getByText('Current Industry: gym')).toBeInTheDocument()
    })
    
    it('shows general markets for unknown industry', () => {
      renderForIndustry(
        <MarketSelector />,
        'general'
      )
      
      expect(screen.getByText('Local Market')).toBeInTheDocument()
      expect(screen.getByText('Regional Market')).toBeInTheDocument()
      expect(screen.getByText('National Market')).toBeInTheDocument()
      expect(screen.getByText('Current Industry: general')).toBeInTheDocument()
    })
  })
  
  describe('Multi-tenant isolation', () => {
    it('preserves market selection across tenant switches', async () => {
      const onMarketChange = jest.fn()
      
      // Render for cinema tenant
      const { rerender, user } = renderForIndustry(
        <MarketSelector onMarketChange={onMarketChange} />,
        'cinema'
      )
      
      // Select cinema market
      const select = screen.getByTestId('market-select')
      await user.selectOptions(select, 'cinema-downtown')
      
      expect(onMarketChange).toHaveBeenCalledWith('cinema-downtown')
      
      // Switch to hotel tenant
      rerender(
        <MarketSelector 
          onMarketChange={onMarketChange} 
          currentMarket="hotel-luxury"
        />,
        { industryType: 'hotel' }
      )
      
      // Should show hotel options now
      expect(screen.getByText('Luxury Hotels')).toBeInTheDocument()
      expect(screen.getByText('Current Industry: hotel')).toBeInTheDocument()
    })
  })
  
  describe('Accessibility', () => {
    it('has no accessibility violations', async () => {
      const { container } = renderWithProviders(
        <MarketSelector currentMarket="general-local" />
      )
      
      await testAccessibility(container)
    })
    
    it('supports keyboard navigation', async () => {
      const onMarketChange = jest.fn()
      const { user } = renderWithProviders(
        <MarketSelector onMarketChange={onMarketChange} />
      )
      
      const select = screen.getByTestId('market-select')
      
      // Focus and navigate with keyboard
      select.focus()
      await user.keyboard('{ArrowDown}')
      await user.keyboard('{Enter}')
      
      expect(onMarketChange).toHaveBeenCalled()
    })
    
    it('has proper ARIA labels and descriptions', () => {
      renderWithProviders(<MarketSelector />)
      
      const select = screen.getByTestId('market-select')
      expect(select).toBeInTheDocument()
      
      // Should have accessible name
      expect(select).toBeVisible()
    })
  })
  
  describe('Loading states', () => {
    it('handles loading state gracefully', async () => {
      // Mock API delay
      mockApiEndpoints.mockMarketEdge()
      
      const { container } = renderWithProviders(
        <MarketSelector />
      )
      
      await waitForLoadingToFinish(container)
      
      expect(screen.getByTestId('market-selector')).toBeInTheDocument()
    })
  })
  
  describe('Error handling', () => {
    it('displays error state when API fails', async () => {
      // Mock API failure
      mockApiEndpoints.mockMarketEdge()
      
      renderWithProviders(<MarketSelector />)
      
      // Component should still render with fallback options
      expect(screen.getByTestId('market-selector')).toBeInTheDocument()
    })
  })
  
  describe('Performance', () => {
    it('renders efficiently with many market options', async () => {
      const manyMarkets = generateTestData.competitors(50)
      
      const startTime = performance.now()
      
      renderWithProviders(
        <MarketSelector />
      )
      
      const endTime = performance.now()
      const renderTime = endTime - startTime
      
      // Should render in reasonable time (< 100ms)
      expect(renderTime).toBeLessThan(100)
    })
  })
  
  describe('Integration scenarios', () => {
    describe('Cinema industry', () => {
      beforeEach(() => {
        industryTestScenarios.cinema.ticketingSystem()
      })
      
      it('integrates with cinema ticketing data', async () => {
        const onMarketChange = jest.fn()
        const { user } = renderForIndustry(
          <MarketSelector onMarketChange={onMarketChange} />,
          'cinema'
        )
        
        // Select cinema market
        const select = screen.getByTestId('market-select')
        await user.selectOptions(select, 'cinema-downtown')
        
        expect(onMarketChange).toHaveBeenCalledWith('cinema-downtown')
        
        // Should trigger integration with ticketing system
        await waitFor(() => {
          expect(screen.getByText('Current Industry: cinema')).toBeInTheDocument()
        })
      })
    })
    
    describe('Hotel industry', () => {
      beforeEach(() => {
        industryTestScenarios.hotel.realTimePricing()
      })
      
      it('integrates with real-time pricing data', async () => {
        const onMarketChange = jest.fn()
        const { user } = renderForIndustry(
          <MarketSelector onMarketChange={onMarketChange} />,
          'hotel'
        )
        
        const select = screen.getByTestId('market-select')
        await user.selectOptions(select, 'hotel-luxury')
        
        expect(onMarketChange).toHaveBeenCalledWith('hotel-luxury')
      })
    })
    
    describe('Gym industry', () => {
      beforeEach(() => {
        industryTestScenarios.gym.membershipTracking()
      })
      
      it('integrates with membership tracking', async () => {
        const onMarketChange = jest.fn()
        const { user } = renderForIndustry(
          <MarketSelector onMarketChange={onMarketChange} />,
          'gym'
        )
        
        const select = screen.getByTestId('market-select')
        await user.selectOptions(select, 'gym-premium')
        
        expect(onMarketChange).toHaveBeenCalledWith('gym-premium')
      })
    })
  })
  
  describe('Snapshot tests', () => {
    it('matches snapshot for general industry', () => {
      const { container } = renderWithProviders(<MarketSelector />)
      expect(container.firstChild).toMatchSnapshot()
    })
    
    it('matches snapshot for cinema industry', () => {
      const { container } = renderForIndustry(
        <MarketSelector />,
        'cinema'
      )
      expect(container.firstChild).toMatchSnapshot()
    })
    
    it('matches snapshot for hotel industry', () => {
      const { container } = renderForIndustry(
        <MarketSelector />,
        'hotel'
      )
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})