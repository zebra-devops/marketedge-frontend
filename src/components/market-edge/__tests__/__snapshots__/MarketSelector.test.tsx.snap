// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketSelector Component Snapshot tests matches snapshot for cinema industry 1`] = `
<div
  data-testid="market-selector"
>
  <select
    data-testid="market-select"
  >
    <option
      value=""
    >
      Select Market
    </option>
    <option
      value="general-local"
    >
      Local Market
    </option>
    <option
      value="general-regional"
    >
      Regional Market
    </option>
    <option
      value="general-national"
    >
      National Market
    </option>
  </select>
  <div
    data-testid="market-info"
  >
    Current Industry: 
    general
  </div>
</div>
`;

exports[`MarketSelector Component Snapshot tests matches snapshot for general industry 1`] = `
<div
  data-testid="market-selector"
>
  <select
    data-testid="market-select"
  >
    <option
      value=""
    >
      Select Market
    </option>
    <option
      value="general-local"
    >
      Local Market
    </option>
    <option
      value="general-regional"
    >
      Regional Market
    </option>
    <option
      value="general-national"
    >
      National Market
    </option>
  </select>
  <div
    data-testid="market-info"
  >
    Current Industry: 
    general
  </div>
</div>
`;

exports[`MarketSelector Component Snapshot tests matches snapshot for hotel industry 1`] = `
<div
  data-testid="market-selector"
>
  <select
    data-testid="market-select"
  >
    <option
      value=""
    >
      Select Market
    </option>
    <option
      value="general-local"
    >
      Local Market
    </option>
    <option
      value="general-regional"
    >
      Regional Market
    </option>
    <option
      value="general-national"
    >
      National Market
    </option>
  </select>
  <div
    data-testid="market-info"
  >
    Current Industry: 
    general
  </div>
</div>
`;
